lib/plugins/aws/package/compile/events/schedule.js

```diff
@@ -148,69 +192,150 @@ class AwsCompileScheduledEvents {
             for (const ScheduleExpression of ScheduleExpressions) {
               scheduleNumberInFunction++;
 
-              const scheduleLogicalId = this.provider.naming.getScheduleLogicalId(
-                functionName,
-                scheduleNumberInFunction
-              );
-              const lambdaPermissionLogicalId =
-                this.provider.naming.getLambdaSchedulePermissionLogicalId(
+              if (method === METHOD_SCHEDULER) {
+                if (!hasSchedulerEvents) {
+                  hasSchedulerEvents = true;
+
+                  if (!resources.IamRoleLambdaExecution) {
+                    log.info(
+                      `Remember to add required EventBridge Scheduler permissions to your execution role. Documentation: ${style.link(
+                        'https://docs.aws.amazon.com/scheduler/latest/UserGuide/setting-up.html#setting-up-execution-role'
+                      )}`
+                    );
+                  } else {
+                    const lambdaAssumeStatement =
+                      resources.IamRoleLambdaExecution.Properties.AssumeRolePolicyDocument.Statement.find(
+                        (statement) => statement.Principal.Service.includes('lambda.amazonaws.com')
+                      );
+                    if (lambdaAssumeStatement) {
+                      lambdaAssumeStatement.Principal.Service.push('scheduler.amazonaws.com');
+                    }
+
+                    const policyDocumentStatements =
+                      resources.IamRoleLambdaExecution.Properties.Policies[0].PolicyDocument
+                        .Statement;
+
+                    allowedInvokeResources = [];
+
+                    policyDocumentStatements.push({
+                      Effect: 'Allow',
+                      Action: ['lambda:InvokeFunction'],
+                      Resource: allowedInvokeResources,
+                    });
+                  }
+                }
+
+                if (allowedInvokeResources) {
```

Comment:
Resolution of `allowedInvokeResources`, looks quite confusing, why simply not:

```javascript
if (!resources.IamRoleLambdaExecution) {
  ...
} else {
  ...
  policyDocumentStatements.push({
    Effect: 'Allow',
    Action: ['lambda:InvokeFunction'],
    Resource: [
      { 'Fn::Sub': functionArnWithVars },
      { 'Fn::Sub': `${functionArnWithVars}:*` }
    ],
  });
}
```

Reply:
Actually, the context of that block only executes the first time it encounters a scheduler event. That is why pushing of resources have to happen outside of that block.

Reply:
Ah indeed, forgot about that.

I believe then this needs to be handled as we do in other cases, where we add needed IAM configuration after we processed all function and events (ofc assuming that at least one event in this mode was configured)

Reply:
Can you give an example of another case where we do this?